harness/determined/common/api/authentication.py

```diff
@@ -90,13 +94,22 @@ def _init_session(
             session_user = constants.DEFAULT_DETERMINED_USER
         assert session_user is not None
 
+        # Check the token store to determine if session_user has valid login credentials.
+        # If there are no such credentials, token will be set to (or will remain) None.
         token = self.token_store.get_token(session_user)
         if token is not None and not _is_token_valid(self.master_address, token, cert):
             self.token_store.drop_user(session_user)
             token = None
 
+        # Get token for session_user given the following logic:
+        # * If session_user has valid login credentials in token store, then we have found our
+        # user-token pair.
+        # * If there is no valid token for session_user, use token from environment (if it exists)
+        # when session_user is stored in the environment and session_user is either the same as
+        # requested_user (the user passed in with a flag) or when no user was passed in wtih a flag.
```

Comment:
Special case: use the token provided from the container environment if:
- no token was obtained from the token store already,
- no user was explicitly requested, or the requested user matches the token available in the container environment, and
- there is a token available from the container environment.